### PR TITLE
irinterp: Fix accidentally introduced deletion of effectful statement

### PR DESCRIPTION
**Pull request by Keno**
I moved around some code in #49692 that broadened the replacement of statements by their const results. This is fine for how we're currently using irinterp in base, because we're requiring some fairly strong effects, but some downstream pipelines (and potentially Base in the future) want to use irinterp on code with arbitrary effects, so put in an appropriate check.

### CODE REVIEW

File: base/compiler/abstractinterpretation.jl [hunk 1]
```diff
@@ -2544,6 +2544,7 @@ end
 function abstract_eval_statement(interp::AbstractInterpreter, @nospecialize(e), vtypes::VarTable, sv::InferenceState)
     if !isa(e, Expr)
         if isa(e, PhiNode)
+            add_curr_ssaflag!(sv, IR_FLAG_EFFECT_FREE)
```
Comment (aviatesk) [inline base/compiler/abstractinterpretation.jl:2547]: Why is this needed? This function works on `InferenceState` and thus irinterp will never use it. The ordinary inference may see PhiNode somehow?

Comment (Keno) [reply base/compiler/abstractinterpretation.jl:2547]: We made it legal in untyped code, but it is true that we may need to add the flag in ssa conversion also.

Comment (aviatesk) [reply base/compiler/abstractinterpretation.jl:2547]: I wonder if there is any actual users of this `PhiNode` handling within untyped code inference. And it sounds a good idea to add the flag during SSA conversion.

**Review by aviatesk — APPROVED** [review]

File: base/compiler/abstractinterpretation.jl (hunk 1 above)
Comment (Keno) [reply base/compiler/abstractinterpretation.jl:2547]: > I wonder if there is any actual users of this PhiNode handling within untyped code inference.
Diffractor uses it.

> And it sounds a good idea to add the flag during SSA conversion.
I think we already set this since `stmt_effect_flags` knows about it, but we've of course been moving to not calling that function quite as often.
